Support new COOL environment structure

dav3r opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR makes updates in support of the switchover from our legacy account/environment scheme (where staging and production accounts exist within the same AWS organization) to our new scheme (where they do not).

Specifically, this PR modifies the Terraform code that creates the user that runs the Molecule tests for this Ansible role such that the user can only operate in a single environment (e.g. dev, staging, production, etc).

This PR also updates our GitHub Actions workflow so that the Molecule tests run in our development environment; previously they ran in production.

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

This gets us closer to our modernized account scheme where there is no more co-mingling of staging and production accounts. Doing that will result in cleaner code across all COOL-related repositories as well as improved separation of resources across all COOL environments.

Related PRs:
* https://github.com/cisagov/molecule-iam-user-tf-module/pull/61
* https://github.com/cisagov/ci-iam-user-tf-module/pull/54

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

These changes were tested by creating Molecule test users in each of our environments, uploading their respective secrets to their associated GitHub environment, then validating that the tests (in the `build.yml` workflow file) all ran successfully.

* [x] Test these changes in a branch of a descendant of this repository (e.g. [`cisagov/ansible-role-cobalt-strike`](https://github.com/cisagov/ansible-role-cobalt-strike)) and verify that everything works in a "real" (non-skeleton) repository/role.

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been -->
<!-- approved. -->

- [x] Revert dependencies to default branches (`cisagov/molecule-iam-user-tf-module` in `terraform/user.tf`)

dav3r committed:
Modify backend to use a partial configuration

This lets us avoid hardcoding the bucket name and allows it to be set on a per-environment basis via the -backend-config command line option or other methods.  For details, see: https://developer.hashicorp.com/terraform/language/backend#partial-configuration

dav3r committed:
Make terraform_state_bucket a required variable and use it to fetch remote states

Also, make all remote states use the same environment (no more mixed staging and production states).

dav3r committed:
Update providers to use a single environment (no more mixed staging and production)

dav3r committed:
Add *.tfconfig and *.tfvars to .gitignore

.tfconfig is the extension we have chosen to use for our backend config files.

dav3r committed:
Remove production and staging providers when calling cisagov/molecule-iam-user-tf-module

Also, replace an unused SSM parameter with a dummy value.

dav3r committed:
Replace staging- and production-specific role outputs with a single role output

dav3r committed:
Run the test job in the dev-a environment

We plan to come up with a more comprehensive solution that will allow
testing in each environment, but we are not there yet.

dav3r committed:
Update instructions in README for creating test users

jsf9k commented on README.md:
```diff
@@ -16,51 +16,95 @@ create the AWS test user.
-with the new encrypted environment variables. This should be done
-using the
+Once the test user is created you will need to update the
+[repository's secrets](https://help.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets)
```
Suggestion:
```suggestion
[repository's environment secrets](https://help.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets)
```

dav3r replied:
I intentionally didn't mention environment secrets when I did https://github.com/cisagov/skeleton-packer/pull/393 because I felt that folks didn't have to use environment secrets if they didn't want to (or if they just had a single environment).  If you think we should explicitly mention environment secrets, then we should do so in `skeleton-packer` as well to be consistent.

jsf9k commented on README.md:
```diff
@@ -16,51 +16,95 @@ create the AWS test user.
-existing secrets on the [appropriate
-page](https://github.com/cisagov/skeleton-ansible-role-with-test-user/settings/secrets)
+If you have appropriate permissions for the repository you can view existing
+secrets on the
```
Suggestion:
```suggestion
environment secrets on the
```

jsf9k commented on README.md:
```diff
@@ -16,51 +16,95 @@ create the AWS test user.
-page](https://github.com/cisagov/skeleton-ansible-role-with-test-user/settings/secrets)
+If you have appropriate permissions for the repository you can view existing
+secrets on the
+[appropriate page](https://github.com/cisagov/debian-packer/settings/secrets)
```
I think this should reference this skeleton repository:
```suggestion
[appropriate page](https://github.com/cisagov/skeleton-ansible-role-with-test-user/settings/secrets)
```

dav3r replied:
Yes, thanks - that was a copy/paste error by me.

jsf9k commented on .github/workflows/build.yml:
```diff
@@ -170,6 +170,9 @@ jobs:
         uses: mxschmitt/action-tmate@v3
         if: env.RUN_TMATE
   test:
+    # All tests run in the dev-a environment until we figure out a more
+    # comprehensive solution to running tests in multiple environments.
```
I created an issue for this that we should reference here:
```suggestion
    # TODO:  Figure out a plan to test in all environments, not just dev-a.
    # See cisagov/skeleton-ansible-role-with-test-user#183 for more details.
```

jsf9k reviewed (changes requested):
I made a few suggestions and requested a few changes that I think are necessary.

dav3r committed:
Use correct repository in URL

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

dav3r committed:
Add link in comment to an issue for testing in all environments

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

jsf9k reviewed (approved):
Powerful work!

jsf9k committed:
Add example code showing how to pass GHA secrets to Molecule

In many repos that inherit from this skeleton we need to specify a
bucket name where third-party installers, packages, licenses, etc. are
stored.  Previously we made this bucket name a role var with a default
value equal to the name of the bucket in our production COOL.  Now
we have multiple COOL environments, and we eventually want to test in
all environments to ensure that the third-party buckets in all
environments contain the expected objects.

Therefore, it no longer makes sense to use a default value for the
role variable.  Instead we lookup an environment variable in
converge.yml and have the GHA workflow set that variable when running
Molecule.

Co-authored-by: David Redmin <david.redmin@gwe.cisa.dhs.gov>

dv4harr10 commented:
Hi Dave @dav3r , for the ReadMe.md file, the following link returns a 404 error - 'This is not the web page you are looking for': sentence reads as "If you have appropriate permissions for the repository you can view existing secrets on the [appropriate page](https://github.com/cisagov/skeleton-ansible-role-with-test-user/settings/secrets) in the repository's settings.", link is "appropriate page".

dav3r commented:
> Hi Dave @dav3r , for the ReadMe.md file, the following link returns a 404 error - 'This is not the web page you are looking for': sentence reads as "If you have appropriate permissions for the repository you can view existing secrets on the [appropriate page](https://github.com/cisagov/skeleton-ansible-role-with-test-user/settings/secrets) in the repository's settings.", link is "appropriate page".

That is because you don't have the appropriate permissions to view that page, so no cause for concern there.

felddy reviewed (approved):
:shipit:
